netdata/netdata
# eBPF unittest + bug fix

##### Summary
Last week after we merge DEB/RPM plugin PRs, it was discovered an issue with eBPF.plugin when legacy codes were not available. This PR is adding the first unit tests for functions that load these binaries and it is also fixing the reported issue on Slack.

We will continue adding unit tests according we are modifying plugin.

##### Test Plan

1. Compile this branch
2. Run netdata.
3. Stop netdata and move contents inside `/usr/libexec/netdata/plugins.d/ebpf.d/` for another directory.
4. Start netdata again. In this step you should have warnings like this inside your collector log:

```sh
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_hardirq.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_ext4.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_process.5.10.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_softirq.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_sync.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_syncfs.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_msync.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_fsync.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_fdatasync.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_sync_file_range.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_disk.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_oomkill.5.4.o: No such file or directory
/var/log/netdata/collector.log:libbpf: elf: failed to open /usr/libexec/netdata/plugins.d/ebpf.d/pnetdata_ebpf_mdflush.5.4.o: No such file or directory
```
but, plugin won't fail anymore.
5. Stop netdata
6. Return files to `/usr/libexec/netdata/plugins.d/ebpf.d/`.
7. Run manually:
```sh
# ./ebpf.plugin --unittest
```

##### Additional Information

<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? Only when installation fails
- How is the user impacted by the change? We are improving plugin and creating basis to run plugin more stable.
- What are there any benefits of the change? A more stable plugin.
</details>
